feat: retention options to delete old uploads

2color opened the pull request:
## What

This is a new option to delete existing uploads in the Storacha space that the `storacha-proof` has permission for.


## Why

Storacha doesn't deduplicate across uploads. This means that every upload will use up space for the full size of the CAR. If each build is large, this can easily add up to a lot of space.


## Open question

- Since we use the same proof/did for uploads, it may need additional permissions when getting created as we currently recommend in the README.md: `w3 delegation create did:key:DID_OF_KEY -c space/blob/add -c space/index/add -c filecoin/offer -c upload/add --base64`
  - We need to add `space/blob/remove` and `upload/remove` to make it possible.

2color committed:
feat: add option to delete old storacha builds

SgtPooki commented on action.yml:
```diff
@@ -55,6 +55,10 @@ inputs:
   storacha-proof:
     description: 'Storacha Base64 encoded proof UCAN with capabilities for the space `w3 delegation create did:key:DID_OF_KEY -c space/blob/add -c space/index/add -c filecoin/offer -c upload/add --base64`'
     required: false
+  storacha-delete-old-builds:
+    description: 'Delete old builds from Storacha before uploading new one. Use "true" or "false" (as strings). Be careful with this option as it will delete all uploads to the space, not just the ones from this build.'
```
Is there some key/tag with storacha we could use to ensure only certain items are removed?

lidel replied:
We should replace it with something that "keeps N latest items" or "purge items older than T" 
See https://github.com/ipshipyard/ipfs-deploy-action/pull/36#discussion_r2296433269

lidel commented on action.yml:
```diff
@@ -174,6 +178,23 @@ runs:
+        npm install -g @web3-storage/w3cli
+        echo "ℹ️ Deleting old builds from Storacha"
+        w3 space add ${{ inputs.storacha-proof }}
+        w3 ls | while read -r cid; do
+          if [[ -n "$cid" ]]; then
+            echo "Removing $cid"
+            w3 rm "$cid"
```
### Problem

I worry this will lead to race conditions in default state where people use our action without going extra mile to set up concurrency limits in their workflow.

Running on multiple PRs during busy day could lead to state where recently published preview on one PR is deleted by build from other PR.

### Fix

I think if we are able to keep last N CIDs, or only delete "older than a week" then we could remove very old ones, but dont risk PRs rug-pulling each other by removing recently created ones.


#### Missing timestamps in `w3` ?

@alanshaw  Looking at sources of  https://github.com/storacha/w3cli / https://github.com/storacha/upload-service the API DOES return timestamps (`insertedAt`, `updatedAt`), but then `w3` CLI  filters them out, even when I pass `w3 ls --json` (line 172-175 in `lib.js` explicitly maps only `{ root, shards }`)

Could pin creation timestamps be exposed, or am I missing something?

We need them to
- identify "old" uploads - timestamps exist in the API but aren't accessible via CLI
- implement "keep last N by age" - no way to sort by time

alanshaw replied:
https://github.com/storacha/upload-service/pull/411

lidel committed:
feat: add retention options for old pins

replaces storacha-delete-old-builds with backend-agnostic retention options:
- retention-days: remove pins older than N days
- retention-count: keep only N most recent pins
- retention-max-removals: safety limit (default 3) to prevent mass deletion
- retention-dry-run: preview deletions without removing

key improvements:
- works with all backends (storacha, ipfs cluster, pinata, kubo, filebase)
- cleanup happens after successful upload for safety
- both retention policies can be used together
- current cid always protected from removal
- kubo: embeds timestamps in pin names for full retention support
- kubo: detects version compatibility (requires v0.29.0+ for named pins)
- clear progress indicators and removal limits

replaces @web3-storage/w3cli with @storacha/cli

note: this implementation has not been tested with real backends and needs
real-world validation before merging

addresses feedback from #36

lidel commented:
@2color switched to latest storacha cli with support for timestamps Alan shipped (https://github.com/ipfs/ipfs-deploy-action/pull/36#discussion_r2321266771) , and pushed https://github.com/ipfs/ipfs-deploy-action/pull/36/commits/07c9786c620cec27ed38ec1d56ed31c506ef8fa4 with prototype that is backend-agnostic:

> 
> replaces storacha-delete-old-builds with backend-agnostic retention options:
> - retention-days: remove pins older than N days
> - retention-count: keep only N most recent pins
> - retention-max-removals: safety limit (default 3) to prevent mass deletion
> - retention-dry-run: preview deletions without removing
> 
> key improvements:
> - works with all backends (storacha, ipfs cluster, pinata, kubo, filebase)
> - cleanup happens after successful upload for safety
> - both retention policies can be used together
> - current cid always protected from removal
> - kubo: embeds timestamps in pin names for full retention support
> - kubo: detects version compatibility (requires v0.29.0+ for named pins)
> 
> replaces @web3-storage/w3cli with @storacha/cli

next step would be to test this PR in some disposable repo with every backend.
if we dont have bandwidth for that, we could wrap untested ones with

```diff
+            echo "::warning::Retention policies for Kubo backend are experimental and disabled for safety. Manual cleanup required."
+            echo "⚠️ Skipping retention management for this backend (not yet tested)"
+            # TODO: Enable after testing
+            if false; then
// code
+            fi
```

2color commented on action.yml:
```diff
@@ -174,20 +188,174 @@ runs:
+            echo "::error::Custom pin-name is not compatible with retention options. Retention policies require the default naming pattern to identify related pins."
+            exit 1
+          fi
+          # Custom pin name for all backends
```
This comment can probably be removed

2color replied:
Or rather, document here why we need a separate pin name for Kubo

2color committed:
fix: use correct env var for storacha cli

2color committed:
fix: link input docs

2color committed:
fix: quote the ucan proof

2color committed:
docs: fix missing capability needed for storacha retention

2color committed:
fix: formatting

2color commented:
> I just got confirmation from Storacha that the space/blob/remove and upload/remove capabilities are needed for deletions.

We also need to add the `upload/list` for retention policies to workj. Fixed in https://github.com/ipfs/ipfs-deploy-action/pull/36/commits/6776fc401a4f413b60f4858a1c9d34677b690522

2color committed:
fix: extract the cid from the json output correctly

2color committed:
chore: move env var assignments to step config

2color committed:
fix: bug when dry-run is false non-zero error

lidel committed:
feat: simplify retention to count-based only

- remove retention-days parameter (had date parsing issues)
- keep retention-count for N most recent builds
- add defensive MAX_REMOVALS fallback (default 3)

lidel committed:
fix: storacha retention filter to properly exclude current CID

The previous jq filter compared an object (.root) with a string ($current),
which always evaluated to true since they're different types. This meant
the current CID could appear in the removal candidates list, relying only
on the bash safety check to prevent its deletion.

Changed to .root["/"] != $current to properly compare the actual CID
string value, ensuring the current upload is filtered out at the jq level.

Storacha uses DAG-JSON format where CIDs are stored as {"/": "cid"}.

2color commented:
Triage notes:

- This is still too high risk, because we delete based on APIs outside our control that can change.
- Ideally, this should be implemented by the different backends, in the best case, using two mechanisms:
  - TTL based mechanism: uploads are deleted automatically after a TTL set on upload expires
  - Retention count: for a given namespace/bucket/prefix you only keep the last N uploads
